In [9]:
!pip install git+https://github.com/Abiyzelalem27/bell_state.git

Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
  Cloning https://github.com/Abiyzelalem27/bell_state.git to /tmp/pip-req-build-dxsl0q0k
  Running command git clone --filter=blob:none --quiet https://github.com/Abiyzelalem27/bell_state.git /tmp/pip-req-build-dxsl0q0k
  Resolved https://github.com/Abiyzelalem27/bell_state.git to commit 74519f7f9f8c21e4e1eb6c12eca3e13015b424a7
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done


In [1]:
import sys
!{sys.executable} -m pip install numpy scipy matplotlib qutip

Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels


In [33]:
import numpy as np
import qutip as qt

from qutip import Bloch
from scipy.linalg import expm, eigh

import Comp_Quant_Dynam as cqd   # custom package
import bell_state as bs          # custom package
from Comp_Quant_Dynam.operators import X, Y, Z, I

# Computational Quantum Dynamics  
## Lecture 1

# 1.1 Recap: Postulates of Quantum Mechanics

## 1. States
The state of a physical system is completely specified by a vector in a Hilbert space:

$$
|\psi\rangle \in \mathcal{H},
\qquad
\langle \psi | \psi \rangle = 1
$$

In [34]:
psi = 1/2 * np.array([1, 0]) + (np.sqrt(3)/2) * np.array([0, 1])
print("psi =", psi)
norm = np.vdot(psi, psi)
print("normalized=", norm)

psi = [0.5       0.8660254]
normalized= 0.9999999999999999


In [35]:
Psi = 1/2 * np.array([1, 0]) + 1/2 * np.sqrt(3) * np.array([0, 1])
print("psi =", psi)
norm = np.vdot(psi, psi)
print("normalized=", norm)

psi = [0.5       0.8660254]
normalized= 0.9999999999999999


In [36]:
psi = np.array([1/2, np.sqrt(3)/2])
print(psi)
norm = np.vdot(psi, psi)
print(norm)

[0.5       0.8660254]
0.9999999999999999


# 2. Operators (Observables in Quantum Mechanics)

Observable physical quantities are represented by **linear Hermitian operators** acting on the Hilbert space.

An operator acts on a quantum state and transforms it into another quantum state.

$$
A : \mathcal{H} \rightarrow \mathcal{H}
$$

meaning that the operator \(A\) maps vectors from the Hilbert space \(\mathcal{H}\) back into the same Hilbert space.

Its action on a state is written as

$$
|\psi\rangle \mapsto A|\psi\rangle
$$

where:

- \( |\psi\rangle \) is the initial quantum state
- \( A|\psi\rangle \) is the transformed state

For observables in quantum mechanics, the operator must satisfy the Hermitian condition

$$
A^{\dagger} = A
$$

which guarantees that measurement outcomes (eigenvalues) are real numbers.

## Examples of Observable Operators

- Position operator:
  
  $$
  \hat{x}
  $$

- Momentum operator:
  
  $$
  \hat{p}
  $$

- Hamiltonian (energy):
  
  $$
  \hat{H}
  $$

- Spin operators:
  
  $$
  S_x,\; S_y,\; S_z
  $$

# Example: Operator Changing a Quantum State

Consider the qubit state

$$
|\psi\rangle =
\begin{pmatrix}
1 \\
0
\end{pmatrix}
=
|0\rangle
$$

Now apply the Pauli-X operator

$$
X =
\begin{pmatrix}
0 & 1 \\
1 & 0
\end{pmatrix}
$$

The action of the operator on the state is

$$
X|0\rangle
=
\begin{pmatrix}
0 & 1 \\
1 & 0
\end{pmatrix}
\begin{pmatrix}
1 \\
0
\end{pmatrix}
=
\begin{pmatrix}
0 \\
1
\end{pmatrix}
=
|1\rangle
$$

Thus, the operator transforms the quantum state

$$
|0\rangle \rightarrow |1\rangle
$$

The Pauli-X operator acts like a quantum NOT gate, flipping the qubit state.

In [65]:
theta = np.pi / 2  
axis = [1, 0, 0] 
Rx = bs.operators.rotation_gate(theta, axis)
#the quantum state
psi = np.array([1, 0], dtype=complex)
print("Initial state=", psi)
# Apply the operator to the quantum state
final_state = Rx @ psi
print("Final state=", final_state)

Initial state= [1.+0.j 0.+0.j]
Final state= [0.70710678+0.j         0.        -0.70710678j]


# 3. Measurements

The possible outcomes of a measurement of an observable \(A\) are the eigenvalues \(a\), defined through the eigenvalue equation

$$
A|a\rangle = a|a\rangle,
\qquad a \in \mathbb{R}
$$

where:

- \(A\) is a Hermitian operator (observable)
- \(a\) is a measurable eigenvalue
- \(|a\rangle\) is the corresponding eigenstate

When a measurement is performed, the quantum state collapses into one of the eigenstates of the observable.

After the measurement, the system is found in the state

$$
|a\rangle
$$

This process is called a **projective measurement**.

The probability of obtaining the measurement outcome \(a\) is given by the **Born rule**

$$
P(a) = |\langle a | \psi \rangle|^2
$$

where:

- $$ \|\psi\rangle\ $$is the initial quantum state
- $$ \|a\rangle\ $$is the eigenstate associated with eigenvalue \(a\)

The Born rule gives the probability of collapsing into the state  $$ \|a\rangle\  $$.

In [66]:
psi = (1 / np.sqrt(2)) * np.array([1, 1])
print("psi=", psi)
#basis states
zero = np.array([1, 0])   
one  = np.array([0, 1])   
#using Born rule for calculated probabilities
P0 = np.abs(np.vdot(zero, psi))**2
P1 = np.abs(np.vdot(one, psi))**2
print("P(0)=", P0)
print("P(1)=", P1)
# Simulate a measurement
measurement = np.random.choice(["|0>", "|1>"], p=[P0, P1])
print("Measurement outcome=", measurement)

psi= [0.70710678 0.70710678]
P(0)= 0.4999999999999999
P(1)= 0.4999999999999999
Measurement outcome= |1>


## 4. Probabilities
The probability of obtaining outcome \(a\) is

$$
|\langle a | \psi \rangle|^2
$$

Hence, the average outcome (expectation value) of \(A\) is

$$
\langle A \rangle
=
\langle \psi | A | \psi \rangle
$$

In [67]:
theta = np.pi / 2  
axis = [0, 0, 1] 
Rz = bs.operators.rotation_gate(theta, axis)
psi = (1 / np.sqrt(2)) * np.array([1, 1])

# Compute expectation value <psi|Z|psi> and  <psi|Rz|psi> to see the differnt
expectation_value_Z = np.vdot(psi, Z @ psi)
print("expec__value_Z=", expectation_value_Z)
expectation_value_Rz = np.vdot(psi, Rz @ psi)
print("expec_value_Rz=", expectation_value_Rz)

expec__value_Z= 0j
expec_value_Rz= (0.7071067811865475+0j)


In [68]:
theta = np.pi / 2  
axis = [1, 0, 0] 
Rx = bs.operators.rotation_gate(theta, axis)
psi = (1 / np.sqrt(2)) * np.array([1, 1])

# Compute expectation value <psi|X|psi> and  <psi|Rx|psi> to see the differnt
expectation_value_X = np.vdot(psi, X @ psi)
print("expec__value_X=", expectation_value_X)
expectation_value_Rx = np.vdot(psi, Rx @ psi)
print("expec_value_Rx=", expectation_value_Rx)

expec__value_X= (0.9999999999999998+0j)
expec_value_Rx= (0.7071067811865475-0.7071067811865474j)


In [69]:
theta = np.pi / 2  
axis = [0, 1, 0] 
Ry = bs.operators.rotation_gate(theta, axis)
psi = (1 / np.sqrt(2)) * np.array([1, 1])

# Compute expectation value <psi|Y|psi> and  <psi|Ry|psi> to see the differnt
expectation_value_Y = np.vdot(psi, Y @ psi)
print("expec__value_Y=", expectation_value_Y)
expectation_value_Ry = np.vdot(psi, Ry @ psi)
print("expec_value_Ry=", expectation_value_Ry)

expec__value_Y= 0j
expec_value_Ry= (0.7071067811865475+0j)


In [70]:
theta = 2*np.pi   
axis = [0, 0, 1] 
R0 = bs.operators.rotation_gate(theta, axis)
psi = (1 / np.sqrt(2)) * np.array([1, 1])

# Compute expectation value <psi|I|psi> and  <psi|R0|psi> to see the differnt
expectation_value_I = np.vdot(psi, I @ psi)
print("expec__value_I=", expectation_value_I)
expectation_value_R0 = np.vdot(psi, R0 @ psi)
print("expec_value_R0=", expectation_value_R0)

expec__value_I= (0.9999999999999998+0j)
expec_value_R0= (-0.9999999999999998+0j)


# Why is 

$$
\langle \psi | Z | \psi \rangle = 0
$$

?

The state

$$
|\psi\rangle
=
\frac{1}{\sqrt{2}}
\left(
|0\rangle + |1\rangle
\right)
$$

contains:

- 50% 

$$
|0\rangle
$$

- 50% 

$$
|1\rangle
$$

The Pauli-Z operator gives:

- 

$$
+1
$$

for

$$
|0\rangle
$$

- 

$$
-1
$$

for

$$
|1\rangle
$$

So the average becomes

$$
\frac{1}{2}(+1)
+
\frac{1}{2}(-1)
=
0
$$

## Simple idea

The positive and negative contributions cancel each other.

---

# Why is 

$$
\langle \psi | R_z | \psi \rangle \neq 0
$$

?

The rotation operator does not measure the state.

Instead, it rotates the quantum state slightly on the Bloch sphere.

After rotation, the new state is still similar to the old state.

So the overlap remains nonzero.

The value

$$
0.707
\approx
\frac{1}{\sqrt{2}}
$$

means:

- the rotated state is not identical to the original state
- but it is still partly aligned with the original state

---

# Simple Physical Picture

## Pauli-Z

Measures:

“Are we in 

$$
|0\rangle
$$

or

$$
|1\rangle
$$

?”

Equal probabilities give average zero.

Equal probabilities give average zero.

---

## Rotation 

$$
R_z
$$

Changes the direction of the quantum state.

Like rotating an arrow slightly:

- it changes direction
- but still points partly in the original direction

So the overlap stays nonzero.

# Conclusion from the Expectation Values

For the quantum state

$$
|\psi\rangle
=
\frac{1}{\sqrt{2}}
\left(
|0\rangle + |1\rangle
\right)
$$

we computed the expectation values of the Pauli operators:

$$
\langle \psi | X | \psi \rangle
$$

$$
\langle \psi | Y | \psi \rangle
$$

$$
\langle \psi | Z | \psi \rangle
$$

The results were approximately:

$$
\langle X \rangle = 1
$$

$$
\langle Y \rangle = 0
$$

$$
\langle Z \rangle = 0
$$

---

# Physical Interpretation

The expectation values determine the direction of the qubit on the Bloch sphere.

Since

$$
\langle X \rangle = 1
$$

and

$$
\langle Y \rangle = 0,
\qquad
\langle Z \rangle = 0
$$

the Bloch vector points completely along the positive \(x\)-direction.

Therefore, the state

$$
|\psi\rangle
=
\frac{1}{\sqrt{2}}
\left(
|0\rangle + |1\rangle
\right)
$$

lies on the positive \(X\)-axis of the Bloch sphere.

---

# Simple Picture

- \(X = 1\) means the qubit points fully along the \(x\)-direction.
- \(Y = 0\) means there is no component along the \(y\)-direction.
- \(Z = 0\) means the qubit is halfway between \( |0\rangle \) and \( |1\rangle \).

So the state points directly along the positive \(x\)-axis.

In [71]:
theta = np.pi / 2  
axis = [0, 0, 1] 
Rz = bs.operators.rotation_gate(theta, axis)
psi = np.array([1, 0])

# Compute expectation value <psi|Z|psi> and  <psi|Rz|psi> to see the differnt
expectation_value_Z = np.vdot(psi, Z @ psi)
print("expec__value_Z=", expectation_value_Z)
expectation_value_Rz = np.vdot(psi, Rz @ psi)
print("expec_value_Rz=", expectation_value_Rz)

expec__value_Z= (1+0j)
expec_value_Rz= (0.7071067811865476-0.7071067811865475j)


In [72]:
theta = np.pi / 2  
axis = [1, 0, 0] 
Rx = bs.operators.rotation_gate(theta, axis)
psi = np.array([1, 0])

# Compute expectation value <psi|X|psi> and  <psi|Rx|psi> to see the differnt
expectation_value_X = np.vdot(psi, X @ psi)
print("expec__value_X=", expectation_value_X)
expectation_value_Rx = np.vdot(psi, Rx @ psi)
print("expec_value_Rx=", expectation_value_Rx)

expec__value_X= 0j
expec_value_Rx= (0.7071067811865476+0j)


In [73]:
theta = np.pi / 2  
axis = [0, 1, 0] 
Ry = bs.operators.rotation_gate(theta, axis)
psi = np.array([1, 0])

# Compute expectation value <psi|Y|psi> and  <psi|Ry|psi> to see the differnt
expectation_value_Y = np.vdot(psi, Y @ psi)
print("expec__value_Y=", expectation_value_Y)
expectation_value_Ry = np.vdot(psi, Ry @ psi)
print("expec_value_Ry=", expectation_value_Ry)

expec__value_Y= 0j
expec_value_Ry= (0.7071067811865476+0j)


In [74]:
theta = 2*np.pi   
axis = [0, 0, 1] 
R0 = bs.operators.rotation_gate(theta, axis)
psi = np.array([1, 0])

# Compute expectation value <psi|I|psi> and  <psi|R0|psi> to see the differnt
expectation_value_I = np.vdot(psi, I @ psi)
print("expec__value_I=", expectation_value_I)
expectation_value_R0 = np.vdot(psi, R0 @ psi)
print("expec_value_R0=", expectation_value_R0)

expec__value_I= (1+0j)
expec_value_R0= (-1-1.2246467991473532e-16j)


# Conclusion from the New Results

The quantum state is

$$
|\psi\rangle
=
|0\rangle
=
\begin{pmatrix}
1 \\
0
\end{pmatrix}
$$

We computed:

$$
\langle \psi | Z | \psi \rangle = 1
$$

$$
\langle \psi | X | \psi \rangle = 0
$$

$$
\langle \psi | Y | \psi \rangle = 0
$$

---

# Physical Interpretation

Since

$$
\langle Z \rangle = 1
$$

and

$$
\langle X \rangle = 0,
\qquad
\langle Y \rangle = 0
$$

the Bloch vector points completely along the positive \(z\)-direction.

Therefore, the state

$$
|0\rangle
$$

lies at the north pole of the Bloch sphere.

---

# Why are the Rotation Overlaps Nonzero?

The quantities

$$
\langle \psi | R_x | \psi \rangle
$$

$$
\langle \psi | R_y | \psi \rangle
$$

$$
\langle \psi | R_z | \psi \rangle
$$

measure the overlap between:

- the original state
- the rotated state

The values are approximately

$$
0.707
=
\frac{1}{\sqrt2}
$$

because a rotation by

$$
\theta = \frac{\pi}{2}
$$

changes the state partially, but not completely.

So the rotated state is still partly aligned with the original state.

---

# Simple Picture

- \(Z=1\) means the state points fully along the \(z\)-axis.
- \(X=0\) means no component along \(x\).
- \(Y=0\) means no component along \(y\).

So the qubit points straight upward on the Bloch sphere.

## 5. Dynamics
A system initially in state

$$
|\psi(t=0)\rangle = |\psi_0\rangle
$$

evolves in time according to the Schrödinger equation:

$$
i\hbar\frac{\partial}{\partial t}|\psi(t)\rangle
=
H|\psi(t)\rangle
$$

where \(H\) is the Hamiltonian operator corresponding to the total energy of the system.

---

## 6. Symmetry
The total wave function of a system of identical particles satisfies:

- **Fermions:** antisymmetric under particle exchange
- **Bosons:** symmetric under particle exchange

This becomes relevant when dealing with many indistinguishable particles.

---

## Note
In this lecture, we consider only **non-relativistic quantum mechanics**.

In the relativistic case, the Schrödinger equation is replaced by equations such as:

- the Dirac equation
- the Klein–Gordon equation

However, many of the mathematical methods used to solve the Schrödinger equation also apply to these relativistic equations.

## 1.2 Single particle in 1D

### classically: 
point particle moving in a potential  
$$H = \frac{p^2}{2m} + V(x)$$

**equations of motion:** Hamilton's equations

---

### quantum mechanically: 
$|\psi\rangle$ encodes the probability amplitude for finding the particle at position $x$.

* **Hilbert space:** space of square integrable functions $\mathcal{L}_2$  
  $\psi(x)$, $2\times$ cont. differentiable and $\int |\psi(x)|^2 dx < \infty$  
  $$\psi : \mathbb{R} \rightarrow \mathbb{C} \, , \quad x \mapsto \psi(x)$$

* **scalar product:** $$\langle \psi | \phi \rangle = \int \psi^*(x) \phi(x) dx$$

* **operators:** $$\hat{x} : \psi(x) \rightarrow x \psi(x)$$  
  $$\hat{p} : \psi(x) \rightarrow -i\hbar \partial_x \psi(x)$$

$$\Rightarrow \hat{H} = -\frac{\hbar^2}{2m} \partial_x^2 + V(x) \qquad \text{(time independent for now)}$$


### Schrödinger equation:
$$\left[ i\hbar \partial_t \psi(x,t) = \left[ -\frac{\hbar^2}{2m}\partial_x^2 + V(x) \right] \psi(x,t) \right] \qquad (1)$$

* **initial condition:** $\quad \psi(x, t=0) = \psi_0(x)$
* **boundary condition:** $\quad \psi(x = \pm\infty, t) = 0$  
  *(bound particle, localized)* $\quad \psi'(x = \pm\infty, t) = 0$
* *(other boundary conditions possible, e.g. periodic)*

---

### time independent $H$ $\Rightarrow$ formal solution of SE
$$|\psi(t)\rangle = e^{-i\hat{H}t/\hbar} |\psi_0\rangle$$

concretely for $(1)$, $t$- and $x$-dependent operators appear only in separate terms $\rightarrow$ make **Ansatz**
$$\psi(x,t) = e^{-iEt/\hbar} \phi(x) \quad \text{into } (1)$$

$$\Rightarrow \quad i\hbar \left( -iE/\hbar \right) e^{-iEt/\hbar} \phi(x) = e^{-iEt/\hbar} \hat{H}\phi(x)$$

$$\Leftrightarrow \quad \left[ \hat{H}\phi(x) = E\phi(x) \right] \quad \begin{aligned} &\text{eigenvalue problem} \\ &\text{(stationary Schrödinger eq.)} \end{aligned}$$

---

**eigenstates** $\phi_k(x)$ of $\hat{H}$ are "stationary" solutions of SE:
$$\psi_k(x,t) = e^{-iE_k t/\hbar} \phi_k(x)$$

$\Rightarrow$ The solution of the t-dep. SE for initial state $\psi_0(x)$ is obtained by decomposing into eigenstates
$$\psi_0(x) = \sum_k c_k \phi_k(x) \quad \text{where } c_k = \int \phi_k^*(x)\psi_0(x)dx$$
*(eigenstates are orthonormal!)*

$$\Rightarrow \quad \left\| \psi(x,t) = \sum_k c_k e^{-iE_k t/\hbar} \phi_k(x) \right\| \quad \text{solves the SE.}$$

### Example: Harmonic oscillator

$$V(x) = \frac{1}{2} m \omega^2 x^2$$

Choose appropriate units: energy in $\hbar\omega$, time in $\omega^{-1}$  
$\text{length in } x_{\text{ho}} = \sqrt{\frac{\hbar}{m\omega}}$

In these units:
$$\hat{H} = -\frac{1}{2}\partial_x^2 + \frac{1}{2}x^2 \qquad (2)$$

Eigenvalue problem (stationary SE) can be solved analytically:

* **Eigen-energies** $\quad E_n = \frac{1}{2} + n \quad (n = 0, 1, 2, \dots)$
* **Eigen-states** $\quad \phi_n(x) = \frac{1}{\sqrt{2^n n! \sqrt{\pi}}} e^{-x^2/2} H_n(x)$

with **Hermite polynomials** $$H_0(x) = 1$$
$$H_1(x) = x$$
$$H_2(x) = x^2 - 1 \quad \dots$$

---

**Observation**: There are infinitely many eigenstates!  
*(spectrum is unbounded, $x$ is continuous)* *How to represent this on a computer?!*

* $\Rightarrow$ Need to truncate Hilbert space (at some $n = n_{\text{max}}$), or equivalently, represent $\psi(x)$ on a finite spatial grid.
* $\Rightarrow$ Numerical solutions are always **<u>approximations</u>**!

## 1.3 Spatial discretization, finite differences

define grid of size $L$ and spacing $\Delta x$

$$\begin{aligned}
\xrightarrow{\quad \Delta x \quad | \quad \dots \quad | \quad \dots \quad} \\
-L/2 \qquad\qquad 0 \qquad\qquad L/2 \quad x
\end{aligned}$$

* **# grid points:** $\frac{L}{\Delta x} + 1 = N$ , grid points $x_k \quad (k = 1 \dots N)$
* represent $\psi(x)$ by its values at the grid points  
  $$\psi_k = \psi(x_k) \in \mathbb{C}^N$$
* **goal:** write stationary SE in discretized form  
  $$H \vec{\psi} = E \vec{\psi} \quad \text{where } H \text{ is a matrix } \in \mathbb{C}^{N \times N}$$

---

### • potential term : 
$V(x)\psi(x) \rightarrow V(x_k)\psi(x_k) \Rightarrow$ **diagonal matrix**

$$\begin{pmatrix}
V(x_1) & 0 & \dots & \dots & 0 \\
0 & V(x_2) & & & \vdots \\
\vdots & & \ddots & & \vdots \\
\vdots & & & \ddots & 0 \\
0 & \dots & \dots & 0 & V(x_N)
\end{pmatrix}$$

---

### • kinetic energy part : 
$\frac{\partial^2}{\partial x^2} \psi(x) \rightarrow$ approximate the derivative on the grid

recall how derivative at $x$ is defined:
$$\begin{aligned}
\frac{\partial \psi}{\partial x} &= \lim_{\Delta x \to 0} \frac{\psi(x) - \psi(x - \Delta x)}{\Delta x} \\
&= \lim_{\Delta x \to 0} \frac{\psi(x + \Delta x) - \psi(x)}{\Delta x}
\end{aligned}$$

$\Rightarrow$ **on the grid, take**
$$\left. \frac{\partial \psi}{\partial x} \right|_{x=x_k} \approx \frac{\psi_k - \psi_{k-1}}{\Delta x} \approx \frac{\psi_{k+1} - \psi_k}{\Delta x}$$

For the 2nd derivative, we can simply iterate this "finite difference" approximation:

$$\begin{aligned}
\left. \frac{\partial^2 \psi}{\partial x^2} \right|_{x=x_k} &= \left. \frac{\partial}{\partial x} \underbrace{\left( \frac{\partial}{\partial x} \psi(x) \right)}_{\psi'} \right|_{x=x_k} = \frac{\psi'_k - \psi'_{k-1}}{\Delta x} \\
&= \frac{\frac{\psi_{k+1} - \psi_k}{\Delta x} - \left( \frac{\psi_k - \psi_{k-1}}{\Delta x} \right)}{\Delta x} = \frac{\psi_{k+1} + \psi_{k-1} - 2\psi_k}{\Delta x^2}
\end{aligned}$$

---

> **Note**: More generally, one can derive such finite difference approximations by interpolating $n$ grid points with an order $n-1$ polynomial and using the derivative of that polynomial at $x_k$.
> 
> The case above corresponds to $n=3$:  
> Quadratic function interpolating points $a_1, a_2, a_3$:
> $$P_3(x) = \frac{(x-a_2)(x-a_3)}{(a_1-a_2)(a_1-a_3)} f(a_1) + \frac{(x-a_1)(x-a_3)}{(a_2-a_1)(a_2-a_3)} f(a_2) + \frac{(x-a_1)(x-a_2)}{(a_3-a_1)(a_3-a_2)} f(a_3)$$
> 
> *check:* $P_3(a_k) = f(a_k)$
> 
> **2nd derivative:**
> $$\begin{aligned}
\left. \frac{\partial^2 P_3}{\partial x^2} \right|_{x=a_2} &= \frac{2}{(-\Delta x)(-2\Delta x)} f(a_1) + \frac{2}{\Delta x(-\Delta x)} f(a_2) + \frac{2}{(2\Delta x)(\Delta x)} f(a_3) \\
&= \frac{f(a_1) + f(a_3) - 2f(a_2)}{\Delta x^2}
\end{aligned}$$

---

$\Rightarrow$ **result: kin. energy term in discrete form:**

$$-\frac{1}{2} \partial_x^2 \psi(x) \longrightarrow -\frac{1}{2\Delta x^2} \begin{pmatrix}
-2 & 1 & 0 & \dots & \dots \\
1 & -2 & 1 & \ddots & \\
0 & 1 & -2 & \ddots & 0 \\
\vdots & \ddots & \ddots & \ddots & 1 \\
\dots & \dots & 0 & 1 & -2
\end{pmatrix}$$

**Summary:** The discretized form of the Hamiltonian $(2)$ is:

$$H = \begin{pmatrix}
\frac{1}{\Delta x^2} + V(x_1) & -\frac{1}{2\Delta x^2} & 0 & \dots & \dots \\
-\frac{1}{2\Delta x^2} & \frac{1}{\Delta x^2} + V(x_2) & \ddots & \ddots & \vdots \\
0 & \ddots & \ddots & \ddots & 0 \\
\vdots & \ddots & \ddots & \ddots & -\frac{1}{2\Delta x^2} \\
\dots & \dots & 0 & -\frac{1}{2\Delta x^2} & \frac{1}{\Delta x^2} + V(x_N)
\end{pmatrix}$$

---

### Remarks:

* This matrix can be diagonalized numerically using standard methods ($\rightarrow$ exercised lecture 2).
* The precision of the obtained solution will depend on the choice of the spatial grid.  
  *(How to choose reasonably?)*
* $H$ is a **real symmetric** matrix.
* Eigenvectors are only defined up to a complex multiplicative constant *(keep in mind for numerics!)*.